airbyte-cdk/python/unit_tests/sources/file_based/stream/concurrent/test_file_based_concurrent_cursor.py

```diff
@@ -0,0 +1,449 @@
+            ("newfile.csv", "2021-01-05T00:00:00.000000Z"),
+            {"newfile.csv": "2021-01-05T00:00:00.000000Z"},
+            [("pending.csv", "2020-01-05T00:00:00.000000Z")],
+            "2020-01-05T00:00:00.000000Z_pending.csv",
+            id="add-to-empty-history-pending-file-is-older",
+        ),
+        pytest.param(
+            {"history": {}},
+            [("newfile.csv", "2021-01-05T00:00:00.000000Z"), ("pending.csv", "2022-01-05T00:00:00.000000Z")],
+            ("newfile.csv", "2021-01-05T00:00:00.000000Z"),
+            {"newfile.csv": "2021-01-05T00:00:00.000000Z"},
+            [("pending.csv", "2022-01-05T00:00:00.000000Z")],
+            "2022-01-05T00:00:00.000000Z_pending.csv",
+            id="add-to-empty-history-pending-file-is-newer",
+        ),
+        pytest.param(
+            {"history": {"existing.csv": "2021-01-04T00:00:00.000000Z"}},
+            [("newfile.csv", "2021-01-05T00:00:00.000000Z")],
+            ("newfile.csv", "2021-01-05T00:00:00.000000Z"),
+            {"existing.csv": "2021-01-04T00:00:00.000000Z", "newfile.csv": "2021-01-05T00:00:00.000000Z"},
+            [],
+            "2021-01-05T00:00:00.000000Z_newfile.csv",
+            id="add-to-nonempty-history-single-pending-file",
+        ),
+        pytest.param(
+            {"history": {"existing.csv": "2021-01-04T00:00:00.000000Z"}},
+            [("newfile.csv", "2021-01-05T00:00:00.000000Z"), ("pending.csv", "2020-01-05T00:00:00.000000Z")],
+            ("newfile.csv", "2021-01-05T00:00:00.000000Z"),
+            {"existing.csv": "2021-01-04T00:00:00.000000Z", "newfile.csv": "2021-01-05T00:00:00.000000Z"},
+            [("pending.csv", "2020-01-05T00:00:00.000000Z")],
+            "2020-01-05T00:00:00.000000Z_pending.csv",
+            id="add-to-nonempty-history-pending-file-is-older",
+        ),
+        pytest.param(
+            {"history": {"existing.csv": "2021-01-04T00:00:00.000000Z"}},
+            [("newfile.csv", "2021-01-05T00:00:00.000000Z"), ("pending.csv", "2022-01-05T00:00:00.000000Z")],
+            ("newfile.csv", "2021-01-05T00:00:00.000000Z"),
+            {"existing.csv": "2021-01-04T00:00:00.000000Z", "newfile.csv": "2021-01-05T00:00:00.000000Z"},
+            [("pending.csv", "2022-01-05T00:00:00.000000Z")],
+            "2022-01-05T00:00:00.000000Z_pending.csv",
+            id="add-to-nonempty-history-pending-file-is-newer",
+        ),
+    ]
+)
+def test_add_file(
+    initial_state: MutableMapping[str, Any],
+    pending_files: List[Tuple[str, str]],
+    file_to_add: Tuple[str, str],
+    expected_history: Dict[str, Any],
+    expected_pending_files: List[Tuple[str, str]],
+    expected_cursor_value: str,
+):
+    cursor = _make_cursor(initial_state)
+    cursor._pending_files = {uri: RemoteFile(uri=uri, last_modified=datetime.strptime(timestamp, DATE_TIME_FORMAT)) for uri, timestamp in pending_files}
```

Comment:
If it represents the user usage, yes! I would also challenge the assertions though as the user would probably not call `cursor._pending_files` as well

Reply:
> I would also challenge the assertions though as the user would probably not call cursor._pending_files as well

True, but I think I'm okay with that in this case because the assertion is there to ensure that the internal state has been set up appropriately when `add_file` is called.